internal/stanza/emitter.go

```diff
@@ -17,46 +17,182 @@ package stanza // import "github.com/open-telemetry/opentelemetry-collector-cont
+	return logEmitterOptionFunc(func(le *LogEmitter) {
+		le.maxBatchSize = maxBatchSize
+		le.batch = make([]*entry.Entry, 0, maxBatchSize)
+	})
+}
+
+// LogEmitterWithFlushInterval returns an option that makes the LogEmitter use the specified flush interval
+func LogEmitterWithFlushInterval(flushInterval time.Duration) LogEmitterOption {
+	return logEmitterOptionFunc(func(le *LogEmitter) {
+		le.flushInterval = flushInterval
+	})
+}
+
+// LogEmitterWithLogger returns an option that makes the LogEmitter use the specified logger
+func LogEmitterWithLogger(logger *zap.SugaredLogger) LogEmitterOption {
+	return logEmitterOptionFunc(func(le *LogEmitter) {
+		le.OutputOperator.BasicOperator.SugaredLogger = logger
+	})
 }
 
 // NewLogEmitter creates a new receiver output
-func NewLogEmitter(logger *zap.SugaredLogger) *LogEmitter {
-	return &LogEmitter{
+func NewLogEmitter(opts ...LogEmitterOption) *LogEmitter {
+	le := &LogEmitter{
 		OutputOperator: helper.OutputOperator{
 			BasicOperator: helper.BasicOperator{
 				OperatorID:    "log_emitter",
 				OperatorType:  "log_emitter",
-				SugaredLogger: logger,
+				SugaredLogger: zap.NewNop().Sugar(),
 			},
 		},
-		logChan: make(chan *entry.Entry),
+		logChan:       make(chan []*entry.Entry),
+		maxBatchSize:  defaultMaxBatchSize,
+		batch:         make([]*entry.Entry, 0, defaultMaxBatchSize),
+		flushInterval: defaultFlushInterval,
 	}
+
+	for _, opt := range opts {
+		opt.Apply(le)
+	}
+
+	return le
+}
+
+// Start starts the goroutine(s) required for this operator
+func (e *LogEmitter) Start(_ operator.Persister) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	e.cancel = cancel
+
+	e.wg.Add(1)
+	go e.flusher(ctx)
+	return nil
 }
 
 // Process will emit an entry to the output channel
 func (e *LogEmitter) Process(ctx context.Context, ent *entry.Entry) error {
+	batchToFlush := e.appendEntry(ent)
+	e.flush(ctx, batchToFlush)
+
+	return nil
+}
+
+// appendEntry appends the entry to the current batch. If maxBatchSize is reached, a new batch will be made, and the old batch
+// (which should be flushed) will be returned
+func (e *LogEmitter) appendEntry(ent *entry.Entry) []*entry.Entry {
+	e.batchMux.Lock()
+	defer e.batchMux.Unlock()
+
+	e.batch = append(e.batch, ent)
+	if uint(len(e.batch)) >= e.maxBatchSize {
+		// flushTriggerAmount triggers a blocking flush
+		return e.makeNewBatchNoLock()
+	}
+
+	return nil
+}
+
+// flusher flushes the current batch every flush interval. Intended to be run as a goroutine
+func (e *LogEmitter) flusher(ctx context.Context) {
+	defer e.wg.Done()
+
+	ticker := time.NewTicker(e.flushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			batch := e.makeNewBatch()
+			e.flush(ctx, batch)
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// flush flushes the provided batch to the log channel. If the batch is nil, does nothing
+func (e *LogEmitter) flush(ctx context.Context, batch []*entry.Entry) {
+	if batch == nil {
+		return
+	}
```

Comment:
In my opinion, we should just not call this function with a nil batch. (i.e. do the nil check before calling)